---
title: "Untitled"
output: html_document
params:
  project:
    value: 2019-02_leukemia
  experiment_peak:
    value: 2019-02_leukemia
  condition:
    value: control
  path_bed:
    value: ../../../peaks/gene_body/projects/
  ensembl_biotype:
    value: id_biotype.rds
---

<!-- Load libraries -->
```{r}
library(dplyr)
library(tidyr)
library(gdata)
library(rtracklayer)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(GenomicRanges)
library(ggplot2)

source("helper_functions.R")
```

```{r}
peaks <- import(paste(params$path_bed, params$experiment_peak, 
                      "/views/archive/exomePeaks_views/archive/", params$condition, "/peak.bed", sep=""), format="bed")
features <- import("../data/Mus_musculus.GRCm38.91.gff3.gz")
```

```{r}
seqlevels(peaks) <- sub("chr", "", seqlevels(peaks))

transcripts <- subset(features, type %in% c("CDS", "three_prime_UTR", "five_prime_UTR"))
intersect_transcripts <- findOverlaps(peaks, transcripts)
intersect_transcripts <- as.data.frame(intersect_transcripts)
colnames(intersect_transcripts) <- c("index_peaks", "index_transcripts")
```

```{r}
peaks_data       <- as.data.frame(peaks)
peaks_data$index_peaks <- as.integer(row.names(peaks_data))
transcripts_data <- as.data.frame(transcripts)
transcripts_data$index_transcripts <- as.integer(row.names(transcripts_data))
transcripts_data <- transcripts_data %>% dplyr::select("type", "Parent", "index_transcripts") 
```

```{r}
peaks_transcripts_data <- peaks_data %>% left_join(intersect_transcripts) %>% 
                                         left_join(transcripts_data, by = c("index_transcripts")) %>%
                                         dplyr::select(-blocks)
```

```{r}
short <- peaks_transcripts_data %>% 
  dplyr::select(index_peaks, type, Parent) %>% unique()
short$Parent <- gsub("transcript:", "", as.character(short$Parent))
short <- na.omit(short)
```

```{r}
ensembl_biotype <- readRDS(paste("../projects/", params$project, "/data/", 
                                 params$ensembl_biotype, sep=""))

protein_coding <- ensembl_biotype %>% 
  filter(gene_biotype == "protein_coding") %>% dplyr::select(-gene_biotype) %>% 
                 dplyr::rename(Parent = ensembl_transcript_id) %>% unique()
short <- protein_coding %>% left_join(short) 
short <- na.omit(short) %>% ungroup()
```

```{r}
cds   <- short %>% filter(type == "CDS") %>% dplyr::select(-type) %>% unique() 
three <- short %>% filter(type == "three_prime_UTR") %>% dplyr::select(-type) %>% unique() 
five  <- short %>% filter(type == "five_prime_UTR") %>% dplyr::select(-type) %>% unique() 
```

```{r}
cds_three   <- dplyr::intersect(cds, three) %>% dplyr::select(-Parent) %>% unique() %>% mutate(type = "Stop")
cds_five    <- dplyr::intersect(cds, five) %>% dplyr::select(-Parent) %>% unique() %>% mutate(type = "Start")
three_five  <- dplyr::intersect(three, five) %>% dplyr::select(-Parent) %>% unique() 
three_five_cds  <- dplyr::intersect(three, five, cds) %>% dplyr::select(-Parent) %>% unique()
only_three  <- setdiff(three, cds) %>% dplyr::select(-Parent) %>% unique() %>% mutate(type = "3UTR")
only_five   <- setdiff(five, cds) %>% dplyr::select(-Parent) %>% unique()  %>% mutate(type = "5UTR")
only_cds    <- setdiff(cds, three) %>% setdiff(five) %>% dplyr::select(-Parent) %>% unique() %>% mutate(type = "CDS")
```

```{r}
all <- bind_rows(cds_three, cds_five, only_three, only_five, only_cds)
#all <- all %>% group_by(type) %>% summarise(sum = n())
all$type <- factor(all$type, levels=c("3UTR", "Stop", "CDS", "Start", "5UTR"))
```

```{r}
p <- ggplot(all, aes(x=factor(1), fill=type))+
  geom_bar(width = 1)+
  coord_polar("y")

p

ggsave(paste("../projects/",  params$project, "/results/", params$condition, "_pie.pdf", sep="") , p)
```

```{r}

```

```{r eval=FALSE, include=FALSE}
cds_three   <- dplyr::intersect(cds, three)  %>% unique() %>% mutate(type = "Stop")
cds_five    <- dplyr::intersect(cds, five)  %>% unique() %>% mutate(type = "Start")
three_five  <- dplyr::intersect(three, five)  %>% unique() 
three_five_cds  <- dplyr::intersect(three, five, cds) %>% unique() 
only_three  <- setdiff(three, cds)  %>% unique() %>% mutate(type = "3UTR")
only_five   <- setdiff(five, cds)  %>% unique()  %>% mutate(type = "5UTR")
only_cds    <- setdiff(cds, three) %>% setdiff(five) %>%  unique() %>% mutate(type = "CDS")
```
